### ISSUE TITLE
Implementation of text direction from language tags

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-thing-description#1750]
@aphillips this issue/question should probably be answered you, I have failed to understand the TD and Strings on the Web spec while working on the implementation of this.

In the TD spec, we have the following assertion that is currently at risk:

- td-text-direction-language-tag : For the MultiLanguage Map, TD Consumers MAY infer the [=base direction=] from the language tag of the individual strings.

This asks for the TD processors to understand the text direction based on a language tag. However, I have read the https://www.w3.org/International/questions/qa-direction-from-language and it says that we cannot infer it from the language tag. We actually need it for our TD Playground implementation, the specific issue tracking this is https://github.com/thingweb/thingweb-playground/issues/384

We simply need to truncate the text in a UI and need to decide where to put the `...` indicating that the text is truncated. In an LTR text, it goes to the right, in an RTL text it goes to the left. If the TD has `@direction` we can use this to put the `...` in the left or right in the `title` and `description` values. However, this cannot be the full solution since:
- There can be no `@direction` tag
- `titles` and `descriptions` have other languages than the default language of the TD, meaning that the `@direction` cannot be used for these.

Some solutions I have thought of:
- Having a lookup table with all the language tags and their direction. This is quite some work and actually believed to be not feasible by the document linked above.
- Embedding the direction in the language tag. I am not sure if this is a common thing but I think it is not.
- Understanding the language from the language tag. Somehow find a way to know the `en` is LTR without a lookup table. I do not see a way how this can be done

Basically, we should understand what is really needed before implementing it. Also, I think that this is not needed in browser-based implementations since the browser engines seem to pick the correct direction when displaying the actual text to the user and I would not see a big problem by leaving this at risk.

**Comment by egekorkan** [comment]
Thank you @aphillips for the explanation, especially the last paragraph really helps. Indeed per-string solution is what we really need due to the multiple languages in TDs.

I will get back to you after our implementation findings. We are using simple JS in that example but I think the idea in your last paragraph is very promising.

**Comment by egekorkan** [comment]
@aphillips I have given another thought and I have another question. We will try the per-string solution as mentioned above however the assertion we need to implement actually asks for figuring out the direction from the language tag. I think that your proposal that we look for `Suppress-Script` makes sense. However, is there another lookup document that tells that `Cyrl` is LTR and `Arab` is RTL? I have found this: https://www.w3.org/International/questions/qa-scripts but it is not using the same notation, e.g. `Arabic` is used instead of `Arab`

I have also found this:
https://en.wikipedia.org/wiki/Module:Language/data/iana_suppressed_scripts

**Comment by aphillips** [comment]
@egekorkan I think it would be better if TD changed the spec :-). Ideally the fallback would be:

* String-specific metadata
* File-level metadata (currently the `@direction` and `@language`)
* (optional) guess from language
* first-strong (i.e. using `dir=auto` in HTML or `FSI`/`PDI` controls around the string)

But I get it. Your question is basically: yes, but... if I want to do the "optional guess" how *do* we guess the direction from the language? 

The `Suppress-Script` method is insufficient for most languages in the LSTR, but "most" is misleading (there are over 7000 primary language subtags, but most of these represent languages with very small speaker populations or which are unwritten). Where information is not available, first-strong should be used (let the text tell you). All strings should be isolated anyway.

The most prevalent mechanism for guessing the script comes from Unicode's [CLDR](https://cldr.unicode.org/) (Common Locale Data Repository) project. The core specification of CLDR is UTR#35 ("LDML"), which contains an algorithm for looking up the script subtag for any given language tag. This is called [addLikelySubtags](http://www.unicode.org/reports/tr35/#Likely_Subtags) and you can follow the link to see the description.

In terms of implementations, there is one in the ICU open source libraries from Unicode. ICU is available in Java and C/C++ and there is a new version "ICU4X" that is written in Rust and provides a basic set of functionality that can be used by JS implementations via web assembly. The function in question is [here](https://icu4x.unicode.org/doc/icu/locid_transform/provider/struct.LikelySubtagsV1.html). ECMA-402 may add "likely subtag" support to `Intl.Locale` in the future (I'm not sure of the status of this).

In ICU4J/ICU4C (but not yet ICU4X) there is a method that lets you ask the guessed direction directly from a language tag ([isRightToLeft](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/ULocale.html#isRightToLeft--)). 

You could build a lookup table directly from CLDR data I think, but I'm not quite sure where that data lives in the vast mess. Searching for how ICU handles it, I found that they have a hardcoded table 😞 about [here](https://github.com/unicode-org/icu/blob/3654e945b68d5042cbf6254dd559a7ba794a76b3/icu4j/main/classes/core/src/com/ibm/icu/lang/UScript.java#L1337) (which you can scan the comment for script codes and the `RTL` value being `|`-ed into the data table. It's probably also somewhere in the [data files](http://unicode.org/Public/cldr/42/)...

If you took the data from UScript (grep for RTL and just make a list of those) you'd be pretty close. What I would actually do is only assert direction for the RTL languages. That is, I'd return a direction of RTL for languages like `ar`, `fa`, `he`, and so on that are written in scripts like `Arab` or `Hebr`, but default to first-strong for non-RTL languages.

**Comment by egekorkan** [comment]
We have implemented it in Playground now: https://github.com/w3c/wot-testing/pull/487 . Can be closed

We will however add above list of steps as informative text (assertions already explain this):

-     String-specific metadata
-     File-level metadata (currently the @direction and @language)
-     (optional) guess from language
-     first-strong (i.e. using dir=auto in HTML or FSI/PDI controls around the string)
